src/codegen/mod.rs

```diff
@@ -563,131 +765,234 @@ fn gen_par_for(
+    cu::Stmt::Seq (
+        Box::new(body),
+        Box::new(sync_stmt),
+    )
+}
+
+fn build_seq(mut stmts: Vec<cu::Stmt>) -> cu::Stmt {
+    if stmts.len() <= 1 {
+        panic!("needs at least two stmts");
+    } else if stmts.len() == 2 {
+        cu::Stmt::Seq(
+            Box::new(stmts.remove(0)),
+            Box::new(stmts.remove(0)),
+        )
+    } else {
+        cu::Stmt::Seq(
+            Box::new(stmts.remove(0)),
+            Box::new(build_seq(stmts)),
+        )
+    }
+}
+
+fn gen_checked_stmt(
+    expr: &desc::Expr,
+    view_ctx: &mut HashMap<String, ViewExpr>,
+    comp_unit: &[desc::FunDef],
+    label: Option<String>,
+) -> cu::Stmt {
+    use desc::ExprKind::*;
+    match &expr.expr {
+        Let(_, _, _, expr1, expr2) => cu::Stmt::Seq(
+            Box::new(gen_checked_stmt(expr1, view_ctx, comp_unit, label.clone())),
+            Box::new(gen_checked_stmt(expr2, view_ctx, comp_unit, label)),
+        ),
+        //gen_checked_stmt(expr1, view_ctx, comp_unit, label), //TODO call on other expr of let
+        Seq(e1, e2) => {
+            cu::Stmt::Seq(
+                Box::new(gen_checked_stmt(e1, view_ctx, comp_unit, label.clone())),
+                Box::new(gen_checked_stmt(e2, view_ctx, comp_unit, label))
+            )
+        }
+        Index(pl_expr, i) => {
+            match label {
+                Some(l) => {
+                    let n = match &pl_expr.ty.as_ref().expect(&format!("{:?}", pl_expr)).ty {
+                        TyKind::Data(DataTy::Array(_, m)) => m,
+                        TyKind::Data(DataTy::Ref(_, _, _, a)) => if let DataTy::Array(_, m) = a.as_ref() {
+                            m
+                        } else {
+                            panic!("cannot index into non array type!");
+                        }, 
+                        TyKind::View(ViewTy::Array(_, m)) => m,
+                        t => panic!("cannot index into non array type!{:?}", t),
+                    };
+                    use crate::ast::*;
+                    cu::Stmt::If{
+                        cond: cu::Expr::BinOp{
+                            op: cu::BinOp::Gt, 
+                            lhs: Box::new(cu::Expr::Nat(i.clone())),
+                            rhs: Box::new(cu::Expr::Nat(n.clone())),
+                        } ,
+                        body: Box::new(cu::Stmt::Block(Box::new(cu::Stmt::Expr(
+                            cu::Expr::Ident(format!("goto {}", l))
+                        )))) ,
+                    }
+                },
+                None => {
+                    cu::Stmt::Expr(cu::Expr::Empty)
+                }
+            }
+           
+        }
+        Assign(_, expr) => gen_checked_stmt(expr, view_ctx, comp_unit, label),
+        App(_, _, _) | ParFor(_, _, _) | PlaceExpr(_) | Lit(_) => cu::Stmt::Expr(cu::Expr::Empty),
+        _ => panic!("Smth not yet implemented {:?}", &expr), // TODO not all cases implemented
+    }
+}
+
+enum CheckedExpr {
```

Comment:
Can this type definition be moved to right before it is used first?

Reply:
done